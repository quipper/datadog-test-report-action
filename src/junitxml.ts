import assert from 'assert'
import * as core from '@actions/core'
import * as fs from 'fs/promises'
import * as path from 'path'
import { XMLParser } from 'fast-xml-parser'

export type TestReport = {
  testFiles: TestFile[]
  testCases: TestCase[]
}

export type TestFile = {
  filename: string
  owners: string[]
  totalTime: number
  totalTestCases: number
}

export type FindOwners = (filename: string) => string[]

export const parseTestReportFiles = async (testReportFiles: string[], findOwners: FindOwners): Promise<TestReport> => {
  const junitXmls = await parseTestReportFilesToJunitXml(testReportFiles)
  const allTestCases: TestCase[] = []
  for (const junitXml of junitXmls) {
    const testCases = findTestCasesFromJunitXml(junitXml, findOwners)
    allTestCases.push(...testCases)
  }
  core.info(`Found ${allTestCases.length} test cases in the test reports`)
  const testFiles = groupTestCasesByTestFile(allTestCases)
  return {
    testFiles,
    testCases: allTestCases,
  }
}

export const groupTestCasesByTestFile = (testCases: TestCase[]): TestFile[] => {
  const testFiles = new Map<string, TestFile>()
  for (const testCase of testCases) {
    const currentTestFile = testFiles.get(testCase.filename) ?? {
      filename: testCase.filename,
      owners: testCase.owners,
      totalTime: 0,
      totalTestCases: 0,
    }
    currentTestFile.totalTime += testCase.time
    currentTestFile.totalTestCases++
    testFiles.set(testCase.filename, currentTestFile)
  }
  return [...testFiles.values()]
}

export type TestCase = {
  name: string
  filename: string
  owners: string[]
  time: number
  success: boolean
}

export const findTestCasesFromJunitXml = (junitXml: JunitXml, findOwners: FindOwners): TestCase[] => {
  const root = junitXml.testsuites?.testsuite ?? junitXml.testsuite ?? []

  function* visit(testSuite: JunitXmlTestSuite): Generator<TestCase> {
    const determineTestCaseFilename = (junitXmlTestCase: JunitXmlTestCase): string => {
      if (junitXmlTestCase['@_file']) {
        return junitXmlTestCase['@_file']
      }
      // For Mocha or Cypress, the first <testsuite> element has the filename of the root suite.
      const mochaRootSuiteFilename = root.at(0)?.['@_file']
      if (mochaRootSuiteFilename) {
        return mochaRootSuiteFilename
      }
      throw new Error(`Element <testcase> must have "file" attribute (name=${junitXmlTestCase['@_name']})`)
    }

    for (const junitXmlTestCase of testSuite.testcase ?? []) {
      const filename = path.normalize(determineTestCaseFilename(junitXmlTestCase))
      yield {
        name: junitXmlTestCase['@_name'],
        filename,
        owners: findOwners(filename),
        time: junitXmlTestCase['@_time'],
        success: !junitXmlTestCase.failure && !junitXmlTestCase.error,
      }
    }
    for (const nestedTestSuite of testSuite.testsuite ?? []) {
      visit(nestedTestSuite)
    }
  }

  const testCases: TestCase[] = []
  for (const testSuite of root) {
    for (const testCase of visit(testSuite)) {
      testCases.push(testCase)
    }
  }
  return testCases
}

const parseTestReportFilesToJunitXml = async (testReportFiles: string[]): Promise<JunitXml[]> => {
  const junitXmls: JunitXml[] = []
  core.startGroup(`Parsing ${testReportFiles.length} test report files`)
  for (const testReportFile of testReportFiles) {
    core.info(`Parsing the test report: ${testReportFile}`)
    const xml = await fs.readFile(testReportFile)
    const junitXml = parseJunitXml(xml)
    junitXmls.push(junitXml)
  }
  core.endGroup()
  return junitXmls
}

type JunitXml = {
  testsuites?: {
    testsuite?: JunitXmlTestSuite[]
  }
  testsuite?: JunitXmlTestSuite[]
}

function assertJunitXml(x: unknown): asserts x is JunitXml {
  assert(typeof x === 'object', 'Root document must be an object')
  assert(x != null, 'Root document must not be null')

  if ('testsuites' in x) {
    assert(typeof x.testsuites === 'object', 'Element <testsuites> must be an object')
    assert(x.testsuites != null, 'Element <testsuites> must not be null')

    if ('testsuite' in x.testsuites) {
      assert(Array.isArray(x.testsuites.testsuite), 'Element <testsuite> must be an array')
      for (const testsuite of x.testsuites.testsuite) {
        assertTestSuite(testsuite)
      }
    }
  }

  if ('testsuite' in x) {
    assert(Array.isArray(x.testsuite), 'Element <testsuite> must be an array')
    for (const testsuite of x.testsuite) {
      assertTestSuite(testsuite)
    }
  }
}

type JunitXmlTestSuite = {
  testsuite?: JunitXmlTestSuite[]
  testcase?: JunitXmlTestCase[]
  '@_file'?: string
}

function assertTestSuite(x: unknown): asserts x is JunitXmlTestSuite {
  assert(typeof x === 'object', 'Element <testsuite> must be an object')
  assert(x != null, 'Element <testsuite> must not be null')
  if ('testsuite' in x) {
    assert(Array.isArray(x.testsuite), 'Element <testsuite> must be an array')
    for (const testsuite of x.testsuite) {
      assertTestSuite(testsuite)
    }
  }
  if ('testcase' in x) {
    assert(Array.isArray(x.testcase), 'Element <testcase> must be an array')
    for (const testcase of x.testcase) {
      assertTestCase(testcase)
    }
  }
}

type JunitXmlTestCase = {
  '@_name': string
  '@_time': number
  '@_file'?: string
  failure?:
    | string
    | {
        '@_message'?: string
      }
  error?:
    | string
    | {
        '@_message'?: string
      }
}

function assertTestCase(x: unknown): asserts x is JunitXmlTestCase {
  assert(typeof x === 'object', `Element <testcase> must be an object but was ${typeof x}`)
  assert(x != null, 'Element <testcase> must not be null')
  assert('@_name' in x, 'Element <testcase> must have "name" attribute')
  assert(typeof x['@_name'] === 'string', `name attribute of <testcase> must be a string but was ${typeof x['@_name']}`)
  assert('@_time' in x, 'Element <testcase> must have "time" attribute')
  assert(typeof x['@_time'] === 'number', `time attribute of <testcase> must be a number but was ${typeof x['@_time']}`)
  if ('@_file' in x) {
    assert(
      typeof x['@_file'] === 'string',
      `file attribute of <testcase> must be a string but was ${typeof x['@_file']}`,
    )
  }
  if ('failure' in x) {
    assert(
      typeof x.failure === 'string' || typeof x.failure === 'object',
      `Element <failure> must be a string or an object but was ${typeof x.failure}`,
    )
    if (typeof x.failure === 'object' && x.failure != null) {
      assert('@_message' in x.failure, 'Element <failure> must have "message" attribute')
      assert(
        typeof x.failure['@_message'] === 'string',
        `message attribute of <failure> must be a string but was ${typeof x.failure['@_message']}`,
      )
    }
  }
  if ('error' in x) {
    assert(
      typeof x.error === 'string' || typeof x.error === 'object',
      `Element <error> must be a string or an object but was ${typeof x.error}`,
    )
    if (typeof x.error === 'object' && x.error != null) {
      assert('@_message' in x.error, 'Element <error> must have "message" attribute')
      assert(
        typeof x.error['@_message'] === 'string',
        `message attribute of <error> must be a string but was ${typeof x.error['@_message']}`,
      )
    }
  }
}

export const parseJunitXml = (xml: string | Buffer): JunitXml => {
  const parser = new XMLParser({
    ignoreAttributes: false,
    removeNSPrefix: true,
    isArray: (_: string, jPath: string): boolean => {
      const elementName = jPath.split('.').pop()
      return elementName === 'testsuite' || elementName === 'testcase'
    },
    attributeValueProcessor: (attrName: string, attrValue: string, jPath: string) => {
      const elementName = jPath.split('.').pop()
      if (
        (elementName === 'testsuites' || elementName === 'testsuite' || elementName === 'testcase') &&
        attrName === 'time'
      ) {
        return Number(attrValue)
      }
      return attrValue
    },
  })
  const parsed: unknown = parser.parse(xml)
  assertJunitXml(parsed)
  return parsed
}
