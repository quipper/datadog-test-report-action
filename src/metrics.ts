import * as core from '@actions/core'
import { v1 } from '@datadog/datadog-api-client'
import { JunitXml, TestCase, TestSuite } from './junitxml.js'

export type Metrics = {
  series: v1.Series[]
  distributionPointsSeries: v1.DistributionPointsSeries[]
}

export type Context = {
  prefix: string
  tags: string[]
  timestamp: number
  filterTestCaseSlowerThan: number
  sendTestCaseSuccess: boolean
  sendTestCaseFailure: boolean
}

export const getJunitXmlMetrics = (junitXml: JunitXml, context: Context): Metrics => {
  const testSuites = junitXml.testsuites?.testsuite ?? junitXml.testsuite ?? []
  const metrics: Metrics = {
    series: [],
    distributionPointsSeries: [],
  }
  for (const testSuite of testSuites) {
    const tsm = getTestSuiteMetrics(testSuite, context)
    metrics.series.push(...tsm.series)
    metrics.distributionPointsSeries.push(...tsm.distributionPointsSeries)
  }
  return metrics
}

const getTestSuiteMetrics = (testSuite: TestSuite, context: Context): Metrics => {
  const metrics: Metrics = {
    series: [],
    distributionPointsSeries: [],
  }
  const tags = [...context.tags, `testsuite_name:${testSuite['@_name']}`]

  metrics.series.push({
    metric: `${context.prefix}.testsuite.count`,
    points: [[context.timestamp, 1]],
    type: 'count',
    tags,
  })
  const duration = testSuite['@_time']
  if (duration > 0) {
    metrics.distributionPointsSeries.push({
      metric: `${context.prefix}.testsuite.duration`,
      points: [[context.timestamp, [duration]]],
      tags,
    })
  }

  for (const testCase of testSuite.testcase ?? []) {
    const tcm = getTestCaseMetrics(testCase, context)
    metrics.series.push(...tcm.series)
    metrics.distributionPointsSeries.push(...tcm.distributionPointsSeries)
  }

  for (const childTestSuite of testSuite.testsuite ?? []) {
    const tsm = getTestSuiteMetrics(childTestSuite, context)
    metrics.series.push(...tsm.series)
    metrics.distributionPointsSeries.push(...tsm.distributionPointsSeries)
  }
  return metrics
}

const getTestCaseMetrics = (testCase: TestCase, context: Context): Metrics => {
  const metrics: Metrics = {
    series: [],
    distributionPointsSeries: [],
  }
  const tags = [...context.tags, `testcase_name:${testCase['@_name']}`]
  if (testCase['@_classname']) {
    tags.push(`testcase_classname:${testCase['@_classname']}`)
  }
  if (testCase['@_file']) {
    tags.push(`testcase_file:${testCase['@_file']}`)
  }

  if (!testCase.failure && !testCase.error) {
    if (context.sendTestCaseSuccess) {
      metrics.series.push({
        metric: `${context.prefix}.testcase.success_count`,
        points: [[context.timestamp, 1]],
        type: 'count',
        tags,
      })
    }
  } else {
    core.error(`FAIL: ${testCase['@_name']}`)
    if (context.sendTestCaseFailure) {
      metrics.series.push({
        metric: `${context.prefix}.testcase.failure_count`,
        points: [[context.timestamp, 1]],
        type: 'count',
        tags,
      })
    }
  }

  const duration = testCase['@_time']
  if (duration > context.filterTestCaseSlowerThan) {
    metrics.distributionPointsSeries.push({
      metric: `${context.prefix}.testcase.duration`,
      points: [[context.timestamp, [duration]]],
      tags,
    })
  }

  return metrics
}

export const unixTime = (date: Date): number => Math.floor(date.getTime() / 1000)
