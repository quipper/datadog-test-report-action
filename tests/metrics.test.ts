import { describe, expect, it } from 'vitest'
import { JunitXml } from '../src/junitxml.js'
import { createMatcher } from '../src/codeowners.js'
import { Context, getJunitXmlMetrics } from '../src/metrics.js'

describe('getJunitXmlMetrics', () => {
  it('returns metrics for a valid JUnit XML', () => {
    const junitXml: JunitXml = {
      testsuites: {
        testsuite: [
          {
            '@_name': 'Tests.Registration',
            '@_time': 6.605871,
            testcase: [
              {
                '@_classname': 'Tests.Registration',
                '@_name': 'testCase1',
                '@_time': 2.113871,
              },
              {
                '@_classname': 'Tests.Registration',
                '@_name': 'testCase2',
                '@_time': 1.051,
              },
              {
                '@_classname': 'Tests.Registration',
                '@_name': 'testCase3',
                '@_time': 3.441,
              },
            ],
          },
          {
            testsuite: [
              {
                '@_name': 'Tests.Authentication.Login',
                '@_time': 4.356,
                testcase: [
                  {
                    '@_name': 'testCase4',
                    '@_classname': 'Tests.Authentication.Login',
                    '@_time': 2.244,
                  },
                  {
                    '@_name': 'testCase5',
                    '@_classname': 'Tests.Authentication.Login',
                    '@_time': 0.781,
                  },
                  {
                    '@_name': 'testCase6',
                    '@_classname': 'Tests.Authentication.Login',
                    '@_time': 1.331,
                  },
                ],
              },
            ],
            '@_name': 'Tests.Authentication',
            '@_time': 9.076816,
            testcase: [
              {
                '@_name': 'testCase7',
                '@_classname': 'Tests.Authentication',
                '@_time': 2.508,
              },
              {
                '@_name': 'testCase8',
                '@_classname': 'Tests.Authentication',
                '@_time': 1.230816,
              },
              {
                failure: {
                  '@_message': 'Assertion error message',
                },
                '@_name': 'testCase9',
                '@_classname': 'Tests.Authentication',
                '@_time': 0.982,
              },
            ],
          },
        ],
      },
    }
    const context: Context = {
      prefix: 'testreport',
      tags: ['env:ci'],
      timestamp: 1234567890,
      filterTestCaseSlowerThan: 1,
      sendTestCaseSuccess: true,
      sendTestCaseFailure: true,
      codeownersMatcher: createMatcher(''),
      testCaseBaseDirectory: '',
    }

    const metrics = getJunitXmlMetrics(junitXml, context)
    expect(metrics.series).toMatchSnapshot()
  })
})
