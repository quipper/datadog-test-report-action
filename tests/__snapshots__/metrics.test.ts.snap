// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getJunitXmlMetrics > returns metrics for a valid JUnit XML 1`] = `
[
  {
    "metric": "testreport.testsuite.count",
    "points": [
      [
        1234567890,
        1,
      ],
    ],
    "tags": [
      "env:ci",
      "testsuite_name:Tests.Registration",
    ],
    "type": "count",
  },
  {
    "metric": "testreport.testcase.success_count",
    "points": [
      [
        1234567890,
        1,
      ],
    ],
    "tags": [
      "env:ci",
      "testcase_name:testCase1",
      "testcase_classname:Tests.Registration",
    ],
    "type": "count",
  },
  {
    "metric": "testreport.testcase.success_count",
    "points": [
      [
        1234567890,
        1,
      ],
    ],
    "tags": [
      "env:ci",
      "testcase_name:testCase2",
      "testcase_classname:Tests.Registration",
    ],
    "type": "count",
  },
  {
    "metric": "testreport.testcase.success_count",
    "points": [
      [
        1234567890,
        1,
      ],
    ],
    "tags": [
      "env:ci",
      "testcase_name:testCase3",
      "testcase_classname:Tests.Registration",
    ],
    "type": "count",
  },
  {
    "metric": "testreport.testsuite.count",
    "points": [
      [
        1234567890,
        1,
      ],
    ],
    "tags": [
      "env:ci",
      "testsuite_name:Tests.Authentication",
    ],
    "type": "count",
  },
  {
    "metric": "testreport.testcase.success_count",
    "points": [
      [
        1234567890,
        1,
      ],
    ],
    "tags": [
      "env:ci",
      "testcase_name:testCase7",
      "testcase_classname:Tests.Authentication",
    ],
    "type": "count",
  },
  {
    "metric": "testreport.testcase.success_count",
    "points": [
      [
        1234567890,
        1,
      ],
    ],
    "tags": [
      "env:ci",
      "testcase_name:testCase8",
      "testcase_classname:Tests.Authentication",
    ],
    "type": "count",
  },
  {
    "metric": "testreport.testcase.failure_count",
    "points": [
      [
        1234567890,
        1,
      ],
    ],
    "tags": [
      "env:ci",
      "testcase_name:testCase9",
      "testcase_classname:Tests.Authentication",
    ],
    "type": "count",
  },
  {
    "metric": "testreport.testsuite.count",
    "points": [
      [
        1234567890,
        1,
      ],
    ],
    "tags": [
      "env:ci",
      "testsuite_name:Tests.Authentication.Login",
    ],
    "type": "count",
  },
  {
    "metric": "testreport.testcase.success_count",
    "points": [
      [
        1234567890,
        1,
      ],
    ],
    "tags": [
      "env:ci",
      "testcase_name:testCase4",
      "testcase_classname:Tests.Authentication.Login",
    ],
    "type": "count",
  },
  {
    "metric": "testreport.testcase.success_count",
    "points": [
      [
        1234567890,
        1,
      ],
    ],
    "tags": [
      "env:ci",
      "testcase_name:testCase5",
      "testcase_classname:Tests.Authentication.Login",
    ],
    "type": "count",
  },
  {
    "metric": "testreport.testcase.success_count",
    "points": [
      [
        1234567890,
        1,
      ],
    ],
    "tags": [
      "env:ci",
      "testcase_name:testCase6",
      "testcase_classname:Tests.Authentication.Login",
    ],
    "type": "count",
  },
]
`;
